File: zen-pos-api/app/routers/attendance.py
from datetime import datetime, timezone
from typing import Optional

from fastapi import APIRouter, Depends, Query

from app.dependencies import require_permission
from app.models.attendance import AttendanceRecordDocument
from app.models.user import UserDocument
from app.schemas.attendance import (
    CheckInRequest, CheckOutRequest, AttendanceRecordOut,
    AttendanceReportOut, AttendanceReportSummary, AttendanceReportEntry,
)
from app.services import attendance_service
from app.core.exceptions import NotFoundError

router = APIRouter()


@router.post("/check-in", response_model=AttendanceRecordOut)
async def check_in(body: CheckInRequest):
    """PIN-verified check-in. No auth token required (kiosk endpoint)."""
    record = await attendance_service.check_in(body.user_id, body.pin)
    return await _to_out(record)


@router.post("/check-out", response_model=AttendanceRecordOut)
async def check_out(body: CheckOutRequest):
    """PIN-verified check-out."""
    record = await attendance_service.check_out(body.user_id, body.pin)
    return await _to_out(record)


@router.get("/today", response_model=list[AttendanceRecordOut])
async def today_records(
    location_id: Optional[str] = Query(None, description="Override location filter"),
    current_user: UserDocument = Depends(require_permission("view_attendance")),
):
    effective_location = current_user.location_id or location_id
    records = await attendance_service.get_today_records(location_id=effective_location)
    return [await _to_out(r) for r in records]


@router.get("/status/{user_id}")
async def user_status(user_id: str):
    """Check if a user is currently checked in today (no auth — kiosk endpoint)."""
    user = await UserDocument.get(user_id)
    if not user:
        raise NotFoundError("User not found")
    today = datetime.now(timezone.utc).strftime("%Y-%m-%d")
    record = await AttendanceRecordDocument.find_one(
        AttendanceRecordDocument.user.id == user.id,  # type: ignore[attr-defined]
        AttendanceRecordDocument.date == today,
        AttendanceRecordDocument.status == "active",
    )
    return {"checked_in": record is not None}


@router.get("/user/{user_id}", response_model=list[AttendanceRecordOut],
            dependencies=[Depends(require_permission("view_attendance"))])
async def user_records(user_id: str):
    records = await attendance_service.get_user_records(user_id)
    return [await _to_out(r) for r in records]


@router.get("/report", response_model=AttendanceReportOut,
            dependencies=[Depends(require_permission("view_hr"))])
async def attendance_report(
    start_date: str = Query(..., description="Start date YYYY-MM-DD"),
    end_date: str = Query(..., description="End date YYYY-MM-DD"),
    user_id: Optional[str] = Query(None, description="Filter to a single user"),
):
    """HR attendance report: aggregate per-user records for a date range."""
    query = AttendanceRecordDocument.find(
        AttendanceRecordDocument.date >= start_date,
        AttendanceRecordDocument.date <= end_date,
    )
    if user_id:
        user = await UserDocument.get(user_id)
        if not user:
            raise NotFoundError("User not found")
        query = query.find(AttendanceRecordDocument.user.id == user.id)  # type: ignore[attr-defined]

    records = await query.to_list()

    # Group by user
    by_user: dict[str, list[AttendanceRecordDocument]] = {}
    for r in records:
        await r.fetch_all_links()
        uid = str(r.user.id) if r.user else ""
        by_user.setdefault(uid, []).append(r)

    summaries: list[AttendanceReportSummary] = []
    for uid, recs in by_user.items():
        u = recs[0].user
        entries = [
            AttendanceReportEntry(
                id=str(r.id),
                user_id=uid,
                user_name=u.name if u else "",
                user_image=u.image if u else "",
                date=r.date,
                check_in=r.check_in,
                check_out=r.check_out,
                status=r.status,
                hours=r.hours,
                is_late=r.is_late,
                is_early_departure=r.is_early_departure,
                is_overtime=r.is_overtime,
            )
            for r in recs
        ]
        summaries.append(AttendanceReportSummary(
            user_id=uid,
            user_name=u.name if u else "",
            user_image=u.image if u else "",
            total_days=len(recs),
            total_hours=sum(r.hours for r in recs),
            late_count=sum(1 for r in recs if r.is_late),
            early_departure_count=sum(1 for r in recs if r.is_early_departure),
            overtime_count=sum(1 for r in recs if r.is_overtime),
            records=entries,
        ))

    return AttendanceReportOut(start_date=start_date, end_date=end_date, summaries=summaries)


async def _to_out(r: AttendanceRecordDocument) -> AttendanceRecordOut:
    await r.fetch_all_links()
    user = r.user
    return AttendanceRecordOut(
        id=str(r.id),
        user_id=str(user.id) if user else "",
        user_name=user.name if user else "",
        date=r.date,
        check_in=r.check_in,
        check_out=r.check_out,
        status=r.status,
        hours=r.hours,
        is_late=r.is_late,
        is_early_departure=r.is_early_departure,
        is_overtime=r.is_overtime,
    )

File: Zen-pos2/src/api/attendance.ts
import { apiRequest } from './client';

export interface AttendanceRecord {
  id: string;
  userId: string;
  userName: string;
  date: string;
  checkIn?: string;
  checkOut?: string;
  status: 'active' | 'completed';
  hours?: number;
  isLate: boolean;
  isEarlyDeparture: boolean;
  isOvertime: boolean;
}

interface ApiRecord {
  id: string;
  user_id: string;
  user_name: string;
  date: string;
  check_in?: string;
  check_out?: string;
  status: string;
  hours?: number;
  is_late: boolean;
  is_early_departure: boolean;
  is_overtime: boolean;
}

function mapRecord(raw: ApiRecord): AttendanceRecord {
  return {
    id: raw.id,
    userId: raw.user_id,
    userName: raw.user_name,
    date: raw.date,
    checkIn: raw.check_in,
    checkOut: raw.check_out,
    status: raw.status as 'active' | 'completed',
    hours: raw.hours,
    isLate: raw.is_late,
    isEarlyDeparture: raw.is_early_departure,
    isOvertime: raw.is_overtime,
  };
}

export async function checkIn(userId: string, pin: string): Promise<AttendanceRecord> {
  const raw = await apiRequest<ApiRecord>('/attendance/check-in', {
    method: 'POST',
    body: JSON.stringify({ user_id: userId, pin }),
  });
  return mapRecord(raw);
}

export async function checkOut(userId: string, pin: string): Promise<AttendanceRecord> {
  const raw = await apiRequest<ApiRecord>('/attendance/check-out', {
    method: 'POST',
    body: JSON.stringify({ user_id: userId, pin }),
  });
  return mapRecord(raw);
}

/**
 * Force-checkout a user without their PIN.
 * Called by management when closing the register.
 * Returns true if the user was checked out, false if they weren't checked in (204).
 */
export async function forceCheckOut(userId: string): Promise<boolean> {
  // apiRequest returns {} as T on 204 — check for the id field to distinguish
  const result = await apiRequest<Partial<ApiRecord>>(`/attendance/force-checkout/${userId}`, { method: 'POST' });
  return !!(result as any)?.id;
}

export async function getUserStatus(userId: string): Promise<boolean> {
  const data = await apiRequest<{ checked_in: boolean }>(`/attendance/status/${userId}`);
  return data.checked_in;
}

export async function getTodayRecords(): Promise<AttendanceRecord[]> {
  const raw = await apiRequest<ApiRecord[]>('/attendance/today');
  return raw.map(mapRecord);
}

export interface AttendanceReportEntry extends AttendanceRecord {
  userImage: string;
}

export interface AttendanceReportSummary {
  userId: string;
  userName: string;
  userImage: string;
  totalDays: number;
  totalHours: number;
  lateCount: number;
  earlyDepartureCount: number;
  overtimeCount: number;
  records: AttendanceReportEntry[];
}

export interface AttendanceReport {
  startDate: string;
  endDate: string;
  summaries: AttendanceReportSummary[];
}

export async function getReport(startDate: string, endDate: string, userId?: string): Promise<AttendanceReport> {
  let url = `/attendance/report?start_date=${startDate}&end_date=${endDate}`;
  if (userId) url += `&user_id=${userId}`;
  const raw = await apiRequest<{
    start_date: string;
    end_date: string;
    summaries: {
      user_id: string; user_name: string; user_image: string;
      total_days: number; total_hours: number; late_count: number;
      early_departure_count: number; overtime_count: number;
      records: (ApiRecord & { user_image: string })[];
    }[];
  }>(url);
  return {
    startDate: raw.start_date,
    endDate: raw.end_date,
    summaries: raw.summaries.map(s => ({
      userId: s.user_id,
      userName: s.user_name,
      userImage: s.user_image,
      totalDays: s.total_days,
      totalHours: s.total_hours,
      lateCount: s.late_count,
      earlyDepartureCount: s.early_departure_count,
      overtimeCount: s.overtime_count,
      records: s.records.map(r => ({ ...mapRecord(r), userImage: s.user_image })),
    })),
  };
}
